README.md
[![opensource](https://badges.frapsoft.com/os/v1/open-source.png?v=103)](#)
[![Licença](https://img.shields.io/badge/licença-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![R](https://img.shields.io/badge/R-4.0+-green.svg)](https://www.r-project.org/)
[![Estrelas](https://img.shields.io/github/stars/AndreBFarias/RankingCongressistas.svg?style=social)](https://github.com/AndreBFarias/RankingCongressistas/stargazers)
[![Contribuições](https://img.shields.io/badge/contribuições-bem--vindas-brightgreen.svg)](https://github.com/AndreBFarias/RankingCongressistas/issues)

# Ranking Congressistas

![Ícone do Congresso](https://raw.githubusercontent.com/AndreBFarias/Ranking-Congressistas/main/assets/logo.png)

## Descrição

Ferramenta para coletar discursos e projetos do Congresso Brasileiro via API Dados Abertos, processar com regex para temas específicos e gerar rankings de deputados e senadores. Identifica posicionamento pró/contra um tema com score ponderado. Dados da API Dados Abertos (Câmara e Senado).

---

### Pré-requisitos
- R 4.0 ou superior.

 #### Pacotes:
  - tidyverse
  - rvest
  -  RCurl
  -  glue
  - lubridate
  - jsonlite
  - httr
  - textreadr
  - doMC
  - readxl
  - stringr

### Instalação

```bash
# Instale pacotes no R:
install.packages(c("tidyverse", "rvest", "RCurl", "glue", "lubridate", "jsonlite", "httr", "textreadr", "doMC", "readxl", "stringr"))
```
### Uso

- Execute `Coleta_camara.R` para discursos da Câmara.
- Execute `Coleta_senado.R` para discursos do Senado.
  (Aplique regex personalizada para temas pró/contra.)
- Execute `Ranking_Politicos_Deputados.R` e `Ranking_Congresso_Senadores.R` para os scores.


### Dependências
Bibliotecas open source para extração e análise.

### Licença
GPLv3 — Livre para usar, modificar e redistribuir, desde que as versões derivadas mantenham a mesma licença.


CHANGELOG.md
# Changelog

## [Não lançado]

## [1.0.0] - 2022-01-01

### Adicionado
- Coleta de discursos da Câmara via API Dados Abertos
- Coleta de discursos do Senado via API Dados Abertos
- Processamento com regex para identificação de posicionamento
- Ranking de deputados por tema
- Ranking de senadores por tema


Coleta_camara.R
# Ritual de Magia Negra Digital: Coleta_Camara_Discursos.R - Invocação das vozes ocultas da Câmara, extraindo discursos como ecos de um coven parlamentar.
# Explicação do Ritual: Coleta_Camara_Discursos.R.md

# 1. Define o início da legislatura atual (57ª, 2023).
# 2. Calcula intervalos mensais até hoje.
# 3. Ajusta datas de fim para cada mês.
# 4. Formata datas para consulta na API.
# 5. Constrói URLs para listar discursos.
# 6. Lê XML de cada URL.
# 7. Conta sessões por XML.
# 8. Extrai detalhes de cada sessão e discurso.
# 9. Constrói dataframe com dados extraídos.
# 10. Função para decodificar RTF de discursos.
# 11. Gera URLs para teor completo dos discursos.
# 12. Define diretório de trabalho.
# 13. Configura paralelismo com doMC.
# 14. Baixa conteúdo das URLs em paralelo com retries.
# 15. Identifica erros nas respostas.
# 16. Lê XML dos URLs com erro.
# 17. Combina respostas válidas.
# 18. Parseia XML final.
# 19. Extrai teor completo e metadados.
# 20. Salva RDS com base atualizada.

library(tidyverse)
library(rvest)
library(RCurl)
library(glue)
library(lubridate)

# 1
inicio_legislatura <- "01/02/2023" %>% dmy()

# 2
inicio <- interval(inicio_legislatura, today()) %>%
  time_length("month") %>%
  trunc() %>% 
  {inicio_legislatura + months(1:.)}

# 3
fim <- inicio + months(1) - 1
fim[length(fim)] <- today()

# 4
inicio <- format(inicio, "%d/%m/%Y")
fim <- format(fim, "%d/%m/%Y")

# 5
c_url2 <- paste0("https://www.camara.leg.br/sitcamaraws/SessoesReunioes.asmx/ListarDiscursosPlenario?dataIni=", inicio, "&dataFim=", fim, "&codigoSessao=&parteNomeParlamentar=&siglaPartido=&siglaUF=")

# 6
lista_discurso <- c_url2 %>% 
  map(~{
    .x %>%
      read_xml()
  })

# 7
lista_sessoes <- lista_discurso %>% 
  map(~{
    .x %>% 
      xml_children() %>%
      xml_name() %>%
      length()
  }) %>% unlist()

# 8
l <- map2(lista_discurso, lista_sessoes, ~{
  s <- .x
  map(1:.y, ~{
    codSessao <- xml_find_all(s, paste0("//sessao[", .x, "]/codigo")) %>% xml_text(trim = T)
    data <- xml_find_all(s, paste0("//sessao[", .x, "]/data")) %>% xml_text(trim = T)
    numero_sessao <- xml_find_all(s, paste0("//sessao[", .x, "]/numero")) %>% xml_text(trim = T)
    tipo_sessao <- xml_find_all(s, paste0("//sessao[", .x, "]/tipo")) %>% xml_text(trim = T)
    codigo_faseSessao <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/codigo")) %>% xml_text(trim = T)
    descricao_faseSessao <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/descricao")) %>% xml_text(trim = T)
    hora_discurso <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/horaInicioDiscurso")) %>% xml_text(trim = T)
    txtIndexaxao <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/txtIndexacao")) %>% xml_text(trim = T)
    numeroQuarto <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/numeroQuarto")) %>% xml_text(trim = T)
    numeroInsercao <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/numeroInsercao")) %>% xml_text(trim = T)
    sumario <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/sumario")) %>% xml_text(trim = T)
    numero_orador <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/orador/numero")) %>% xml_text(trim = T)
    nome_orador <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/orador/nome")) %>% xml_text(trim = T)
    partido_orador <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/orador/partido")) %>% xml_text(trim = T)
    uf_orador <- xml_find_all(s, paste0("//sessao[", .x, "]/fasesSessao/faseSessao/discursos/discurso/orador/uf")) %>% xml_text(trim = T)
    
    cbind(codSessao, data, numero_sessao, tipo_sessao,
          codigo_faseSessao, descricao_faseSessao,
          hora_discurso, txtIndexaxao, numeroQuarto, numeroInsercao, sumario,
          numero_orador, nome_orador, partido_orador, uf_orador)  
  })
})

# 9
df <- map_dfr(seq_along(lista_sessoes), ~{
  listaDetalhe <- l[[.x]]
  map_dfr(1:length(listaDetalhe), ~{
    as.data.frame(listaDetalhe[[.x]], stringsAsFactors = FALSE)
  })
})

# 10
decode_rtf <- function(txt) {
  txt %>%
    base64Decode %>%
    str_replace_all("\\\\'e3", "ã") %>%
    str_replace_all("\\\\'e1", "á") %>%
    str_replace_all("\\\\'e9", "é") %>%
    str_replace_all("\\\\'e7", "ç") %>%
    str_replace_all("\\\\'ed", "í") %>%
    str_replace_all("\\\\'f3", "ó") %>%
    str_replace_all("\\\\'ea", "ê") %>%
    str_replace_all("\\\\'e0", "à") %>%
    str_replace_all("(\\\\[[:alnum:]']+|[\\r\\n]|^\\{|\\}$)", "") %>%
    str_replace_all("\\{\\{[[:alnum:]; ]+\\}\\}", "") %>%
    str_trim
}

# 11
discursos_url <- glue("https://www.camara.leg.br/SitCamaraWS/SessoesReunioes.asmx/obterInteiroTeorDiscursosPlenario?codSessao={df$codSessao}&numOrador={df$numero_orador}&numQuarto={df$numeroQuarto}&numInsercao={df$numeroInsercao}")

# 12
setwd("~/Desktop/DiscursosCongresso")

# 13
library(doMC)
nr_cores <- 60
registerDoMC(nr_cores)

# 14
res_ds_url <- foreach(ds_url = discursos_url) %dopar% {
  tryCatch({
    curlhand <- getCurlHandle() 
    curlSetOpt(.opts = list(forbid.reuse = 1), curl = curlhand) 
    retry(getURL(ds_url, curl = curlhand), max = 10, init = 0)
  }, error = function(e) {
    tryCatch({
      curlhand <- getCurlHandle() 
      curlSetOpt(.opts = list(forbid.reuse = 1), curl = curlhand) 
      retry(getURL(ds_url, curl = curlhand), max = 10, init = 0)
    }, error = function(e) {
      paste0(ds_url, " ## DOUBLE ERROR")
    })
  })
}

# 15
saida <- unlist(res_ds_url)
erro <- which(!grepl('<?xml version=\"1.0\"', saida))
urls_com_erro <- as.list(discursos_url[erro])

# 16
res_ds_url__erro <- lapply(urls_com_erro, read_xml)

# 17
res_ds_url_final <- c(res_ds_url[-erro], res_ds_url__erro)
res_ds_url_final <- res_ds_url_final[grepl('<?xml version=\"1.0\"', res_ds_url_final)]

# 18
res_ds_url_xml_final <- lapply(res_ds_url_final, read_xml, options = "HUGE")

# 19
inteiro_teor <- map_dfr(1:length(res_ds_url_xml_final), ~{
  l <- res_ds_url_xml_final[[.x]]
  orador <- xml_find_first(l, "//nome") %>% xml_text()
  partido <- xml_find_first(l, "//partido") %>% xml_text()
  uf <- xml_find_first(l, "//uf") %>% xml_text()
  horaInicioDiscurso <- xml_find_all(l, "//horaInicioDiscurso") %>% xml_text()
  inteiro <- xml_find_all(l, "//discursoRTFBase64") %>%
    xml_text() %>% 
    decode_rtf()
  cbind(orador, partido, uf, horaInicioDiscurso, inteiro) %>% as_tibble()
})

# 20
saveRDS(inteiro_teor, 'base_discursos_camara_2023_atual.rds')

# "O poder é como uma sombra: quanto mais se busca, mais ela se alonga." 

Coleta_senado.R
library(tidyverse)
library(purrr)
library(plyr)
library(dplyr)
library(jsonlite)
library(httr)


senadores<-function(legislatura_inicio,legislatura_fim){
  url<-paste0("http://legis.senado.leg.br/dadosabertos/senador/lista/legislatura/",legislatura_inicio,"/",legislatura_fim)
  jsonlite::fromJSON(url)
}

leg_55_56<-senadores(55,56)


codigo_discurso<- leg_55_56 %>% 
  purrr::pluck(1) %>% 
  pluck(4) %>% 
  pluck(1) %>% 
  pluck(1) %>% 
  pull(1)


url_discurso<-paste0("https://legis.senado.leg.br/dadosabertos/senador/",codigo_discurso,"/discursos")

base <- url_discurso %>% map(fromJSON)

b <- base %>%
  modify_depth(3,~pluck(.x,2))

b1<-b %>% 
  map(~pluck(.x,1))

b2<-b1 %>% 
  map(~{
    .x %>% 
      pluck(4) %>% 
      pluck(1) %>% 
      pluck(9)
  })

b3<-map2_dfr(codigo_discurso,b2,~{
  cbind(.x,.y) %>% as_tibble()
  
})

saveRDS(b3,"urls_discursos_senado.rds")

f<-function(x){
  
  a<- read_html(x) %>% 
  html_nodes(xpath="//div[@id='content']") %>%
  html_text()
}

library(textreadr)
library(rvest)

d<-NULL
for(i in 1:length(b3$.y)){
  tryCatch({
  a<-f(b3$.y[[i]]) %>%
    tibble::as_tibble(validate=F) %>% 
    magrittr::set_names("discurso") %>%
    dplyr::mutate(codigo=b3$.x[[i]],
                  url_texto=b3$.y[[i]])
  d<-rbind(d,a)
  },error=function(e){
    
    e
  }, finally = {
    
    next
  })
  
}

d <- d[,c(2,3,1)]
table(d$codigo)

saveRDS(d,"base_discursos_senado.rds")


Ranking_Congresso_Senadores.R
library(readr)
library(readxl)
library(stringr)
library(tidyverse)


discursos_senado <- read_csv("senadores_discursos_2015_2020_regex.csv")
projetos_senado <- read_xlsx("base_projetos_senado_2015_2020_regex.xlsx")

## Agrupamento de discursos por senador

discursos_senado$senador <- str_extract(discursos_senado$titulo,'(?<=Pronunciamento de ).*(?= em)')
discursos_senado <- discursos_senado[,c(6:92)]

discursos_por_senador <- discursos_senado %>%
                         group_by(senador) %>%
                         summarise(across(everything(), list(sum)))

discursos_por_senador$total_discursos <- rowSums(discursos_por_senador[,c(2:ncol(discursos_por_senador))])
discursos_por_senador <- discursos_por_senador[,c("senador","total_discursos")]

## Agrupamento de projetos por senador

table(projetos_senado$AutorPrincipal.IdentificacaoParlamentar.NomeParlamentar)

projetos_senado <- projetos_senado[,c(29,66:151)]

projetos_por_senador <- projetos_senado %>%
                        group_by(AutorPrincipal.IdentificacaoParlamentar.NomeParlamentar) %>%
                        summarise(across(everything(), list(sum)))

projetos_por_senador$total_projetos <- rowSums(projetos_por_senador[,c(2:ncol(projetos_por_senador))])
projetos_por_senador <- projetos_por_senador[,c("AutorPrincipal.IdentificacaoParlamentar.NomeParlamentar","total_projetos")]
colnames(projetos_por_senador)[1] <- "senador"

## Montando o ranking

lista_senadores <- discursos_por_senador$senador
lista_senadores <- append(lista_senadores,projetos_por_senador$senador)
lista_senadores <- unique(lista_senadores)
lista_senadores <- lista_senadores[!is.na(lista_senadores)]

ranking_senadores <- data.frame(senador = lista_senadores)
ranking_senadores <- dplyr::left_join(ranking_senadores,projetos_por_senador,by=c("senador"))
ranking_senadores <- dplyr::left_join(ranking_senadores,discursos_por_senador,by=c("senador"))
ranking_senadores[is.na(ranking_senadores)] <- 0

ranking_senadores <- transform(ranking_senadores, projetos_norm = (total_projetos - min(total_projetos)) / (max(total_projetos) - min(total_projetos)))
ranking_senadores <- transform(ranking_senadores, discursos_norm = (total_discursos - min(total_discursos)) / (max(total_discursos) - min(total_discursos)))

ranking_senadores$projetos_x_discursos <- (ranking_senadores$projetos_norm + ranking_senadores$discursos_norm)/2

ranking_senadores <- ranking_senadores %>% arrange(desc(projetos_x_discursos))

library(writexl)

write_xlsx(ranking_senadores,"ranking_senadores.xlsx")


Ranking_Politicos_Deputados.R
library(readr)
library(readxl)
library(stringr)
library(tidyverse)


#discursos_camara <- read_csv("deputadoes_discursos_2015_2020_regex.csv")
projetos_camara <- read_csv("base_projetos_camara_reduzido_2015_2020_regex.csv")

## Agrupamento de projetos por deputado

#table(projetos_camara$Autor_Tipo)

projetos_camara <- projetos_camara[grepl("Deputado",projetos_camara$Autor_Tipo),]

projetos_camara <- projetos_camara[,c(16,18:103)]
colnames(projetos_camara)[1] <- "deputado"
projetos_camara$deputado <- toupper(projetos_camara$deputado)
projetos_camara <- projetos_camara[!grepl("SENADO",projetos_camara$deputado),]

projetos_camara <- projetos_camara %>% 
                   mutate(deputado = strsplit(as.character(deputado), ", ")) %>% 
                   unnest(deputado)

projetos_camara$deputado <- trimws(projetos_camara$deputado)

projetos_por_deputado <- projetos_camara %>%
  group_by(deputado) %>%
  summarise(across(everything(), list(sum)))

projetos_por_deputado$total_projetos <- rowSums(projetos_por_deputado[,c(2:ncol(projetos_por_deputado))])
projetos_por_deputado <- projetos_por_deputado[,c("deputado","total_projetos")]
